### DISCUSSION TITLE
Problem with setValue (by Id) and get selected value (id) with downshift

### DISCUSSION THREAD

**Discussion by Bartyee** [react-hook-form/react-hook-form#5358] [Q&A] [answered]
```js
import React from "react";
import Downshift from "downshift";

const items = [
  { id: 0, name: "apple" },
  { id: 1, name: "pear" }
];

export default ({ value, onChange }) => (
  <Downshift
    onStateChange={(e) => {
      onChange(e.inputValue || "");
    }}
    // inputValue={value}
    inputValue={items.find((el) => el.id === value)?.name}
    itemToString={(item) => (item ? item.name : "")}
  >
    {({
      getInputProps,
      getItemProps,
      getLabelProps,
      getMenuProps,
      isOpen,
      openMenu,
      inputValue,
      highlightedIndex,
      selectedItem
    }) => (
      <div>
        <label {...getLabelProps()} className="label">
          Downshift
        </label>
        <input
          {...getInputProps({
            onFocus: () => {
              if (!isOpen) {
                openMenu();
              }
            }
          })}
          placeholder="Enter a fruit"
        />
        <ul {...getMenuProps()}>
          {isOpen
            ? items
                // .filter((item) => !inputValue || item.includes(inputValue))
                .filter((item) => !inputValue || item.name.includes(inputValue))
                .map((item, index) => (
                  <li
                    {...getItemProps({
                      key: item.id,
                      index,
                      item,
                      style: {
                        backgroundColor:
                          highlightedIndex === index ? "lightgray" : null,
                        fontWeight: selectedItem === item ? "bold" : "normal"
                      }
                    })}
                  >
                    {item.name}
                  </li>
                ))
            : null}
        </ul>
      </div>
    )}
  </Downshift>
);
```
Im tried to setValue by id. If user want to change value then return ID to controller.
What happened:
A component has changed the controlled prop "inputValue" to be uncontrolled. This prop should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled Downshift element for the lifetime of the component. 
https://codesandbox.io/s/friendly-voice-48sry?file=/src/Downshift.js

Please help me, its very important for my project.

**Comment by thanh-nguyen-95** [comment, marked as answer]
I believe this is an error caused by how you use `downshift` rather than a problem of `react-hook-form`.

I'm not familiar with `downshift` at all but my guess is:

- `inputValue` has a value at first (could be an empty string)
- `inputValue` turned into `undefined`, which indicates the component is uncontrolled

It could be the other way around (uncontrolled -> controlled)

```jsx
<Downshift
  inputValue={items.find((el) => el.id === value)?.name} // the problem is here. if no item is found, this would be undefined
  itemToString={(item) => (item ? item.name : "")}
>
</Downshift>
```

Change it to this and see if it helps:

```jsx
<Downshift
  inputValue={items.find((el) => el.id === value)?.name || ''} // set the 'fallback' value to an empty string in case no item found
  itemToString={(item) => (item ? item.name : "")}
>
</Downshift>
```

Hope this helps. If not, I think you should discuss this in `downshift` repo since it's an error from `downshift` in the first place. Cheers!

**Reply by bluebill1049** [reply]
thank you @thanh-nguyen-95 ❤️

**Reply by Bartyee** [reply]
Thanks @thanh-nguyen-95 for reply and help. Unfortunatelly it doesnt work :/
If I want select value, then nothing appear in input. Also I have problem with setValue, If im trying to set that, then component getting bug with selecting etc. :/ 

Any ideas? Autocompletes are my worst component in my forms ... 👎

**Reply by thanh-nguyen-95** [reply]
I can't help you here since I have never used `downshift` before.

I see in the CodeSandbox that you're using Material UI, why don't you have a look at https://material-ui.com/components/autocomplete/. It fits your needs quite well.

Anyway, this is what I would do if I were you:

- Create a side project, doesn't have to be fancy. Just `downshift` and plain `react` for testing
- Re-create your component with `downshift` and `useState` for the value
- Try to make it work. If not, ask for help in `downshift` repo using that code, which is clean and simple without any side effect
- When everything runs smoothly, integrate `react-hook-form` and see if it causes any problem. If it does, feel free to come back to this thread or create a new discussion. We're more than happy to help you!

Good luck!

**Reply by Bartyee** [reply]
> I can't help you here since I have never used `downshift` before.
> 
> I see in the CodeSandbox that you're using Material UI, why don't you have a look at https://material-ui.com/components/autocomplete/. It fits your needs quite well.
> 
> Anyway, this is what I would do if I were you:
> 
> * Create a side project, doesn't have to be fancy. Just `downshift` and plain `react` for testing
> * Re-create your component with `downshift` and `useState` for the value
> * Try to make it work. If not, ask for help in `downshift` repo using that code, which is clean and simple without any side effect
> * When everything runs smoothly, integrate `react-hook-form` and see if it causes any problem. If it does, feel free to come back to this thread or create a new discussion. We're more than happy to help you!
> 
> Good luck!

Ahh... I can't find any good example for my case with  mui autocomplete. I don't know how I can propely setValue by ID (similiar like my issue here). 
I will check that resolve :)

**Reply by Bartyee** [reply]
@thanh-nguyen-95 @bluebill1049
Hello again. Unfortunately I remove downshift from my project because I cant handle it propery. Autocomplete is my the worst component in appliaction and I spent a lot of time to find good library (mui etc.) which help me to provide support RHF + my problem. Could u help me with that?
Now Im trying last time to use mui autocomplete but I can't set value by ID again. I have brainfuck af and I don't know how to do that... please help :(
```js
<form onSubmit={handleSubmit((data) => console.log(data))}>
      <Controller
        name="country"
        as={
          <Autocomplete
            options={options}
            getOptionLabel={(option) => option.name}
            getOptionSelected={(option, value) => {
              if (!option) return option;
              return option?._id === getOpObj(value)?._id;
            }}
            renderInput={(params) => <TextField {...params} label="Country" />}
          />
        }
        onChange={([, obj]) => {
          const op = getOpObj(obj);
          if (op) return getOpObj(obj)._id;
          return null;
        }}
        control={control}
        defaultValue={null}
      />
      <Button type="submit">Submit</Button>
      <Button onClick={() => set()}>setvalue</Button>
    </form>
```


https://codesandbox.io/s/fancy-hooks-ev78o?file=/src/FormTwo.jsx

**Reply by thanh-nguyen-95** [reply]
A working example: https://codesandbox.io/s/optimistic-haslett-28qk1?file=/src/FormTwo.jsx

I recommend using `render` prop of `Controller` since it provides more control over `as` prop. Everything else is just reading the MUI documentation and follow along.

I don't want to sound rude but your code is all over the place. You clearly don't know what you're doing. My suggestion is to keep things simple, learn the basics, take baby steps and please, I cannot stress this enough: read the documentation (both in the website and in the code) of any libraries or frameworks that you use.

Hope this helps. Take a break if you need to. Good luck!

**Reply by Bartyee** [reply]
> A working example: https://codesandbox.io/s/optimistic-haslett-28qk1?file=/src/FormTwo.jsx
> 
> I recommend using `render` prop of `Controller` since it provides more control over `as` prop. Everything else is just reading the MUI documentation and follow along.
> 
> I don't want to sound rude but your code is all over the place. You clearly don't know what you're doing. My suggestion is to keep things simple, learn the basics, take baby steps and please, I cannot stress this enough: read the documentation (both in the website and in the code) of any libraries or frameworks that you use.
> 
> Hope this helps. Take a break if you need to. Good luck!

Thanks! Its not my code, I found it and wanted to change that.
